crates/pathfinder/src/merkle_tree.rs

```diff
@@ -0,0 +1,1089 @@
+            }
+
+            #[test]
+            fn delete() {
+                let mut conn = rusqlite::Connection::open_in_memory().unwrap();
+                let transaction = conn.transaction().unwrap();
+
+                let key0 = StarkHash::from_hex_str("99cadc82").unwrap();
+                let key1 = StarkHash::from_hex_str("901823").unwrap();
+                let key2 = StarkHash::from_hex_str("8975").unwrap();
+
+                let val0 = StarkHash::from_hex_str("1").unwrap();
+                let val1 = StarkHash::from_hex_str("2").unwrap();
+                let val2 = StarkHash::from_hex_str("3").unwrap();
+
+                let mut uut =
+                    MerkleTree::load("test".to_string(), &transaction, ZERO_HASH).unwrap();
+                uut.set(key0, val0).unwrap();
+                let root0 = uut.commit().unwrap();
+
+                let mut uut = MerkleTree::load("test".to_string(), &transaction, root0).unwrap();
+                uut.set(key1, val1).unwrap();
+                let root1 = uut.commit().unwrap();
+
+                let mut uut = MerkleTree::load("test".to_string(), &transaction, root0).unwrap();
+                uut.set(key2, val2).unwrap();
+                let root2 = uut.commit().unwrap();
+
+                let uut = MerkleTree::load("test".to_string(), &transaction, root1).unwrap();
+                uut.delete().unwrap();
+
+                let uut = MerkleTree::load("test".to_string(), &transaction, root0).unwrap();
+                assert_eq!(uut.get(key0).unwrap(), val0);
+                assert_eq!(uut.get(key1).unwrap(), ZERO_HASH);
+                assert_eq!(uut.get(key2).unwrap(), ZERO_HASH);
+
+                let uut = MerkleTree::load("test".to_string(), &transaction, root2).unwrap();
+                assert_eq!(uut.get(key0).unwrap(), val0);
+                assert_eq!(uut.get(key1).unwrap(), ZERO_HASH);
+                assert_eq!(uut.get(key2).unwrap(), val2);
+            }
+        }
+
+        #[test]
+        fn mulitple_identical_roots() {
+            let mut conn = rusqlite::Connection::open_in_memory().unwrap();
+            let transaction = conn.transaction().unwrap();
+            let mut uut = MerkleTree::load("test".to_string(), &transaction, ZERO_HASH).unwrap();
+
+            let key = StarkHash::from_hex_str("99cadc82").unwrap();
+            let val = StarkHash::from_hex_str("12345678").unwrap();
+            uut.set(key, val).unwrap();
+
+            let root0 = uut.commit().unwrap();
+
+            let uut = MerkleTree::load("test".to_string(), &transaction, root0).unwrap();
+            let root1 = uut.commit().unwrap();
+
+            let uut = MerkleTree::load("test".to_string(), &transaction, root1).unwrap();
+            let root2 = uut.commit().unwrap();
+
+            assert_eq!(root0, root1);
+            assert_eq!(root0, root2);
+
+            let uut = MerkleTree::load("test".to_string(), &transaction, root0).unwrap();
```

Comment:
Is it necessary to load it after every operation?

Reply:
The `commit` (and `delete`) consume `self`. So yes its necessary.

You can argue that one could still use a tree after that, but it made sense to me to disallow using it after either persisting it, or deleting it from storage.

Reply:
Ah missed that. I would assume you could build further onto a tree such that:
- Committing turn the internal tree into an Unresolved hash with new root.
- Deleting turn internal tree into a new empty tree (that you could commit).

But no strong opinions, "reloading" also makes sense.

Reply:
One could do that sure, but the usage at the top end will more closely match opening `A`, mutating, committing and then opening `B` and deleting. So no need to have more.